{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyBigWig\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import deepdish as dd\n",
    "import os\n",
    "import pyfaidx\n",
    "import random\n",
    "import pickle as pkl\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import tensorflow as tf\n",
    "import argparse\n",
    "import json\n",
    "import one_hot as dinuc_shuffle_main\n",
    "from tensorflow.keras.utils import get_custom_objects\n",
    "from tensorflow.keras.models import load_model\n",
    "import tensorflow as tf\n",
    "import os\n",
    "os.environ[\"CUDA_DEVICE_ORDER\"]=\"PCI_BUS_ID\"   # see issue #152\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"]=\"5\"\n",
    "%matplotlib inline\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#regions = \"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/ATAC_PE/GM12878/negatives_data/negatives_with_summit.bed\"\n",
    "#regions = \"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/DNASE_SE/IMR90/negatives_data/negatives_with_summit.bed\"\n",
    "#regions = \"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/ATAC_PE/K562/negatives_data/negatives_with_summit.bed\"\n",
    "regions = \"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/ATAC_PE/IMR90/negatives_data/negatives_with_summit.bed\"\n",
    "\n",
    "genome = \"/mnt/lab_data2/anusri/chrombpnet/reference/hg38.genome.fa\"\n",
    "#model_h5 = \"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/DNASE_PE/K562/nautilus_runs_may18/K562_05.13.2022_bias_128_4_1234_0.5_fold_0/chrombpnet_model/chrombpnet_wo_bias.h5\"\n",
    "#model_h5 = \"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/ATAC_PE/K562/nautilus_runs/K562_02.17.2022_bias_128_4_1234_0.5_fold_0/chrombpnet_model/chrombpnet_wo_bias.h5\"\n",
    "#model_h5=\"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/DNASE_SE/IMR90/nautilus_runs_apr12/IMR90_04.09.2022_bias_128_4_1234_0.8_fold_0/chrombpnet_model/chrombpnet_wo_bias.h5\"\n",
    "model_h5=\"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/ATAC_PE/IMR90/nautilus_runs_apr12/IMR90_04.09.2022_bias_128_4_1234_0.4_fold_0/chrombpnet_model/chrombpnet_wo_bias.h5\"\n",
    "#model_h5 =\"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/DNASE_SE/GM12878/nautilus_runs/GM12878_03.06.2022_bias_128_4_1234_0.8_fold_0/chrombpnet_model/chrombpnet_wo_bias.h5\"\n",
    "#model_h5 = \"/mnt/lab_data2/anusri/chrombpnet/results/chrombpnet/ATAC_PE/GM12878/nautilus_runs/GM12878_03.01.2022_bias_128_4_1234_0.4_fold_0/chrombpnet_model/chrombpnet_wo_bias.h5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def softmax(x, temp=1):\n",
    "    norm_x = x - np.mean(x,axis=1, keepdims=True)\n",
    "    return np.exp(temp*norm_x)/np.sum(np.exp(temp*norm_x), axis=1, keepdims=True)\n",
    "\n",
    "\n",
    "def get_footprint_for_motif(seqs, motif, model, inputlen, batch_size):\n",
    "    '''\n",
    "    Returns footprints for a given motif. Motif is inserted in both the actual sequence and reverse complemented version.\n",
    "    seqs input is already assumed to be one-hot encoded. motif is in sequence format.\n",
    "    '''\n",
    "    midpoint=inputlen//2\n",
    "\n",
    "    w_mot_seqs = seqs.copy()\n",
    "    w_mot_seqs[:, midpoint-len(motif)//2:midpoint-len(motif)//2+len(motif)] =dinuc_shuffle_main.dna_to_one_hot([motif])\n",
    "\n",
    "    # midpoint of motif is the midpoint of sequence\n",
    "    pred_output=model.predict(w_mot_seqs, batch_size=batch_size, verbose=False)\n",
    "    footprint_for_motif_fwd = softmax(pred_output[0])*(np.exp(pred_output[1])-1)\n",
    "\n",
    "    # reverse complement the sequence\n",
    "    w_mot_seqs_revc = w_mot_seqs[:, ::-1, ::-1]\n",
    "    pred_output_rev=model.predict(w_mot_seqs_revc, batch_size=batch_size, verbose=False)\n",
    "    footprint_for_motif_rev = softmax(pred_output_rev[0])*(np.exp(pred_output_rev[1])-1)\n",
    "\n",
    "    # add fwd sequence predictions and reverse sesquence predictions (not we flip the rev predictions)\n",
    "    counts_for_motif = (np.exp(pred_output_rev[1]) - 1 + np.exp(pred_output[1]) - 1)/2\n",
    "    footprint_for_motif_tot = (footprint_for_motif_fwd+footprint_for_motif_rev[:,::-1])/2\n",
    "    footprint_for_motif =  footprint_for_motif_tot / footprint_for_motif_tot.sum(axis=1)[:, np.newaxis]\n",
    "\n",
    "    return footprint_for_motif_tot, counts_for_motif \n",
    "\n",
    "def get_footprint_for_two_motifs(seqs, motifs, model, inputlen, batch_size, spacing):\n",
    "    '''\n",
    "    Returns footprints for a given motif. Motif is inserted in both the actual sequence and reverse complemented version.\n",
    "    seqs input is already assumed to be one-hot encoded. motif is in sequence format.\n",
    "    '''\n",
    "    midpoint=inputlen//2\n",
    "\n",
    "    spacing_per_motif = spacing // 2\n",
    "    \n",
    "    w_mot_seqs = seqs.copy()\n",
    "    \n",
    "    motif = motifs[0]\n",
    "    start = midpoint-(len(motif)//2)\n",
    "    w_mot_seqs[:, start:start+len(motif)] = dinuc_shuffle_main.dna_to_one_hot([motif])\n",
    "    #print(motif,start,start+len(motif))\n",
    "    if spacing > 0:\n",
    "        spacing_per_motif = spacing \n",
    "        motif = motifs[1]\n",
    "        start = start+len(motifs[0])+spacing_per_motif \n",
    "        w_mot_seqs[:, start:start+len(motif)] = dinuc_shuffle_main.dna_to_one_hot([motif])\n",
    "    else:\n",
    "        spacing_per_motif = spacing \n",
    "        motif = motifs[1]\n",
    "        start = start + spacing_per_motif - len(motif)\n",
    "        w_mot_seqs[:, start:start+len(motif)] = dinuc_shuffle_main.dna_to_one_hot([motif])\n",
    "    \n",
    "    #print(motif,start,start+len(motif))\n",
    "    \n",
    "    # midpoint of motif is the midpoint of sequence\n",
    "    pred_output=model.predict(w_mot_seqs, batch_size=batch_size, verbose=False)\n",
    "    footprint_for_motif_fwd = softmax(pred_output[0])*(np.exp(pred_output[1])-1)\n",
    "\n",
    "    # reverse complement the sequence\n",
    "    w_mot_seqs_revc = w_mot_seqs[:, ::-1, ::-1]\n",
    "    pred_output_rev=model.predict(w_mot_seqs_revc, batch_size=batch_size, verbose=False)\n",
    "    footprint_for_motif_rev = softmax(pred_output_rev[0])*(np.exp(pred_output_rev[1])-1)\n",
    "\n",
    "    # add fwd sequence predictions and reverse sesquence predictions (not we flip the rev predictions)\n",
    "    counts_for_motif = (np.exp(pred_output_rev[1]) - 1 + np.exp(pred_output[1]) - 1)/2\n",
    "    footprint_for_motif_tot = (footprint_for_motif_fwd+footprint_for_motif_rev[:,::-1])/2\n",
    "    footprint_for_motif =  footprint_for_motif_tot / footprint_for_motif_tot.sum(axis=1)[:, np.newaxis]\n",
    "\n",
    "    return footprint_for_motif_tot, counts_for_motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_seq(peaks_df, genome, width, shuffle=False):\n",
    "    \"\"\"\n",
    "    fetches sequence from a given genome.\n",
    "    \"\"\"\n",
    "    vals = []\n",
    "\n",
    "    for i, r in peaks_df.iterrows():\n",
    "        sequence = str(genome[r['chr']][(r['start']+r['summit'] - width//2):(r['start'] + r['summit'] + width//2)])\n",
    "        if len(sequence) == width:\n",
    "                vals.append(sequence)\n",
    "\n",
    "    return dinuc_shuffle_main.dna_to_one_hot(vals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:No training configuration found in the save file, so the model was *not* compiled. Compile it manually.\n"
     ]
    }
   ],
   "source": [
    "model=load_model(model_h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "NARROWPEAK_SCHEMA = [\"chr\", \"start\", \"end\", \"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"summit\"]\n",
    "inputlen = 2114\n",
    "regions_df = pd.read_csv(regions, sep='\\t', names=NARROWPEAK_SCHEMA)\n",
    "chroms_to_keep = [\"chr1\"]\n",
    "regions_subsample = regions_df[(regions_df[\"chr\"].isin(chroms_to_keep))].sample(100, random_state=0)\n",
    "genome_fasta = pyfaidx.Fasta(genome)\n",
    "regions_seqs = get_seq(regions_subsample, genome_fasta, inputlen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif =  \"\"\n",
    "batch_size=128\n",
    "full_footprint_control = get_footprint_for_motif(regions_seqs, motif, model, inputlen, batch_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_size=512\n",
    "\n",
    "def simulate_motif_set(motifs, model, regions_seqs):\n",
    "\n",
    "    \n",
    "    #print(motifs[0])\n",
    "    #print(motifs[1])\n",
    "    #motif1_only = get_footprint_for_motif(regions_seqs, motifs[0], model, inputlen, batch_size)[1]\n",
    "    #motif2_only = get_footprint_for_motif(regions_seqs, motifs[1], model, inputlen, batch_size)[1]\n",
    "    #total_sum = motif1_only+motif2_only\n",
    "    \n",
    "    #plt.plot(motif1_only)\n",
    "    #plt.plot(motif2_only)\n",
    "\n",
    "    # -/- orientation\n",
    "    motif1_only = []\n",
    "    motif2_only = []\n",
    "    data_in_spacings = []\n",
    "    for spacing in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,20,40,50,60,70,80,90,100,110,120,130,140,150]:\n",
    "    #for spacing in [4,6]:\n",
    "\n",
    "        puu_runx_footprint = get_footprint_for_two_motifs(regions_seqs, [motifs[0], \"\"], model, inputlen, batch_size, spacing=-1*spacing)\n",
    "        motif1_only.append(puu_runx_footprint)\n",
    "\n",
    "        puu_runx_footprint = get_footprint_for_two_motifs(regions_seqs, [\"\", motifs[1]], model, inputlen, batch_size, spacing=-1*spacing)\n",
    "        motif2_only.append(puu_runx_footprint)\n",
    "\n",
    "        puu_runx_footprint = get_footprint_for_two_motifs(regions_seqs, motifs, model, inputlen, batch_size, spacing=-1*spacing)\n",
    "        data_in_spacings.append(puu_runx_footprint)\n",
    "        \n",
    "    # +/+ orientation\n",
    "    motif1_only_rev = []\n",
    "    motif2_only_rev = []\n",
    "    data_in_spacings_rev = []\n",
    "    for spacing in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,20,40,50,60,70,80,90,100,110,120,130,140,150]:\n",
    "    #for spacing in [4,6]:\n",
    "        puu_runx_footprint = get_footprint_for_two_motifs(regions_seqs, [motifs[0], \"\"], model, inputlen, batch_size, spacing=-1*spacing)\n",
    "        motif1_only_rev.append(puu_runx_footprint)\n",
    "\n",
    "        puu_runx_footprint = get_footprint_for_two_motifs(regions_seqs, [\"\", motifs[1]], model, inputlen, batch_size, spacing=-1*spacing)\n",
    "        motif2_only_rev.append(puu_runx_footprint)\n",
    "    \n",
    "        puu_runx_footprint = get_footprint_for_two_motifs(regions_seqs, motifs, model, inputlen, batch_size, spacing=spacing)\n",
    "        data_in_spacings_rev.append(puu_runx_footprint)\n",
    "        \n",
    "        \n",
    "    #all_counts = [x[1] for x in data_in_spacings]\n",
    "    #all_counts_rev = [x[1] for x in data_in_spacings_rev]\n",
    "\n",
    "    distance = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,20,40,50,60,70,80,90,100,110,120,130,140,150]+[-1,-2,-3,-4,-5,-6,-7,-8,-9,-10,-11,-12,-13,-14,-15,-20,-40,-50,-60,-70,-80,-90,-100,-110,-120,-130,-140,-150]\n",
    "\n",
    "    #total_counts = all_counts_rev+all_counts\n",
    "    #distance = np.array(distance)\n",
    "    #total_counts = np.array(total_counts)\n",
    "    \n",
    "    #coop_score = np.max(total_counts)\n",
    "    #coop_distance = distance[np.argmax(total_counts)]\n",
    "   \n",
    "    #print(motif1_only, motif2_only, total_sum, coop_score, np.max(total_counts)/np.mean(total_counts), coop_distance, total_counts,distance)\n",
    "\n",
    "    return [motif1_only, motif2_only, data_in_spacings, motif1_only_rev, motif2_only_rev, data_in_spacings_rev, distance]\n",
    "\n",
    "    #print(motif1_only, motif2_only, total_sum, coop_score, np.max(total_counts)/np.mean(total_counts), coop_distance)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations = pd.read_csv(\"addons/imr90/imr90.counts.tomtom.motifs_string.tsv\",header=0, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pattern</th>\n",
       "      <th>Num_Seqlets</th>\n",
       "      <th>Label</th>\n",
       "      <th>Match_1</th>\n",
       "      <th>q-value</th>\n",
       "      <th>Match_2</th>\n",
       "      <th>q-value.1</th>\n",
       "      <th>Match_3</th>\n",
       "      <th>q-value.2</th>\n",
       "      <th>Match_4</th>\n",
       "      <th>...</th>\n",
       "      <th>q-value.5</th>\n",
       "      <th>Match_7</th>\n",
       "      <th>q-value.6</th>\n",
       "      <th>Match_8</th>\n",
       "      <th>q-value.7</th>\n",
       "      <th>Match_9</th>\n",
       "      <th>q-value.8</th>\n",
       "      <th>Match_10</th>\n",
       "      <th>q-value.9</th>\n",
       "      <th>string</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>metacluster_0.pattern_0</td>\n",
       "      <td>12878</td>\n",
       "      <td>FOS</td>\n",
       "      <td>FOS_HUMAN.H11MO.0.A</td>\n",
       "      <td>1.022300e-03</td>\n",
       "      <td>FOSL2_MOUSE.H11MO.0.A</td>\n",
       "      <td>1.022300e-03</td>\n",
       "      <td>FOSL1_HUMAN.H11MO.0.A</td>\n",
       "      <td>1.022300e-03</td>\n",
       "      <td>FOSL1_MOUSE.H11MO.0.A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004283</td>\n",
       "      <td>FOSB_HUMAN.H11MO.0.A</td>\n",
       "      <td>0.004283</td>\n",
       "      <td>JUNB_MOUSE.H11MO.0.A</td>\n",
       "      <td>0.004283</td>\n",
       "      <td>JUN_MOUSE.H11MO.0.A</td>\n",
       "      <td>0.004393</td>\n",
       "      <td>JUNB_HUMAN.H11MO.0.A</td>\n",
       "      <td>0.004393</td>\n",
       "      <td>ATGACTCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>metacluster_0.pattern_1</td>\n",
       "      <td>9160</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>CTCF_MA0139.1</td>\n",
       "      <td>1.493190e-12</td>\n",
       "      <td>CTCF_HUMAN.H11MO.0.A</td>\n",
       "      <td>8.496500e-10</td>\n",
       "      <td>CTCF_MOUSE.H11MO.0.A</td>\n",
       "      <td>3.108420e-08</td>\n",
       "      <td>CTCF_C2H2_1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>CTCFL_MA1102.1</td>\n",
       "      <td>0.000237</td>\n",
       "      <td>ZIC2_MOUSE.H11MO.0.C</td>\n",
       "      <td>0.139479</td>\n",
       "      <td>RARA_nuclearreceptor_6</td>\n",
       "      <td>0.166742</td>\n",
       "      <td>ZIC3_HUMAN.H11MO.0.B</td>\n",
       "      <td>0.166742</td>\n",
       "      <td>GCGCCCCCTAGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>metacluster_0.pattern_3</td>\n",
       "      <td>6646</td>\n",
       "      <td>FOXC1</td>\n",
       "      <td>Foxc1.mouse_forkhead_2</td>\n",
       "      <td>3.633420e-03</td>\n",
       "      <td>FOXG1_forkhead_1</td>\n",
       "      <td>6.010030e-03</td>\n",
       "      <td>FOXJ3_forkhead_2</td>\n",
       "      <td>6.010030e-03</td>\n",
       "      <td>FOXL1_MA0033.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006467</td>\n",
       "      <td>Foxj3.mouse_forkhead_3</td>\n",
       "      <td>0.006467</td>\n",
       "      <td>FOXJ2_forkhead_2</td>\n",
       "      <td>0.006791</td>\n",
       "      <td>FOXO4_HUMAN.H11MO.0.C</td>\n",
       "      <td>0.006791</td>\n",
       "      <td>FOXO4_MOUSE.H11MO.0.C</td>\n",
       "      <td>0.006791</td>\n",
       "      <td>TGTTTACAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>metacluster_0.pattern_4</td>\n",
       "      <td>4786</td>\n",
       "      <td>FOSB+JUNB</td>\n",
       "      <td>FOSB+JUNB_MA1135.1</td>\n",
       "      <td>3.365410e-01</td>\n",
       "      <td>FOSL1+JUND_MA1142.1</td>\n",
       "      <td>3.365410e-01</td>\n",
       "      <td>FOSL2+JUNB_MA1138.1</td>\n",
       "      <td>3.365410e-01</td>\n",
       "      <td>FOSL2+JUND_MA1144.1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.336541</td>\n",
       "      <td>JDP2_MA0655.1</td>\n",
       "      <td>0.336541</td>\n",
       "      <td>JDP2_bZIP_3</td>\n",
       "      <td>0.336541</td>\n",
       "      <td>JDP2_bZIP_1</td>\n",
       "      <td>0.336541</td>\n",
       "      <td>FOSL1+JUNB_MA1137.1</td>\n",
       "      <td>0.336541</td>\n",
       "      <td>TTAGTCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>metacluster_0.pattern_5</td>\n",
       "      <td>2157</td>\n",
       "      <td>SP1/KLF</td>\n",
       "      <td>KLF12_HUMAN.H11MO.0.C</td>\n",
       "      <td>9.172640e-05</td>\n",
       "      <td>SP3_HUMAN.H11MO.0.B</td>\n",
       "      <td>9.172640e-05</td>\n",
       "      <td>SP3_MOUSE.H11MO.0.B</td>\n",
       "      <td>9.172640e-05</td>\n",
       "      <td>SP1_HUMAN.H11MO.0.A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000545</td>\n",
       "      <td>KLF3_HUMAN.H11MO.0.B</td>\n",
       "      <td>0.000545</td>\n",
       "      <td>KLF3_MOUSE.H11MO.0.A</td>\n",
       "      <td>0.000545</td>\n",
       "      <td>SP4_HUMAN.H11MO.0.A</td>\n",
       "      <td>0.000663</td>\n",
       "      <td>SP4_MOUSE.H11MO.0.B</td>\n",
       "      <td>0.000663</td>\n",
       "      <td>GGCCCCGCCCCCTCCCCG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Pattern  Num_Seqlets      Label                 Match_1  \\\n",
       "0  metacluster_0.pattern_0        12878        FOS     FOS_HUMAN.H11MO.0.A   \n",
       "1  metacluster_0.pattern_1         9160       CTCF           CTCF_MA0139.1   \n",
       "2  metacluster_0.pattern_3         6646      FOXC1  Foxc1.mouse_forkhead_2   \n",
       "3  metacluster_0.pattern_4         4786  FOSB+JUNB      FOSB+JUNB_MA1135.1   \n",
       "4  metacluster_0.pattern_5         2157    SP1/KLF   KLF12_HUMAN.H11MO.0.C   \n",
       "\n",
       "        q-value                Match_2     q-value.1                Match_3  \\\n",
       "0  1.022300e-03  FOSL2_MOUSE.H11MO.0.A  1.022300e-03  FOSL1_HUMAN.H11MO.0.A   \n",
       "1  1.493190e-12   CTCF_HUMAN.H11MO.0.A  8.496500e-10   CTCF_MOUSE.H11MO.0.A   \n",
       "2  3.633420e-03       FOXG1_forkhead_1  6.010030e-03       FOXJ3_forkhead_2   \n",
       "3  3.365410e-01    FOSL1+JUND_MA1142.1  3.365410e-01    FOSL2+JUNB_MA1138.1   \n",
       "4  9.172640e-05    SP3_HUMAN.H11MO.0.B  9.172640e-05    SP3_MOUSE.H11MO.0.B   \n",
       "\n",
       "      q-value.2                Match_4  ...  q-value.5  \\\n",
       "0  1.022300e-03  FOSL1_MOUSE.H11MO.0.A  ...   0.004283   \n",
       "1  3.108420e-08            CTCF_C2H2_1  ...   0.000024   \n",
       "2  6.010030e-03         FOXL1_MA0033.2  ...   0.006467   \n",
       "3  3.365410e-01    FOSL2+JUND_MA1144.1  ...   0.336541   \n",
       "4  9.172640e-05    SP1_HUMAN.H11MO.0.A  ...   0.000545   \n",
       "\n",
       "                  Match_7  q-value.6               Match_8  q-value.7  \\\n",
       "0    FOSB_HUMAN.H11MO.0.A   0.004283  JUNB_MOUSE.H11MO.0.A   0.004283   \n",
       "1          CTCFL_MA1102.1   0.000237  ZIC2_MOUSE.H11MO.0.C   0.139479   \n",
       "2  Foxj3.mouse_forkhead_3   0.006467      FOXJ2_forkhead_2   0.006791   \n",
       "3           JDP2_MA0655.1   0.336541           JDP2_bZIP_3   0.336541   \n",
       "4    KLF3_HUMAN.H11MO.0.B   0.000545  KLF3_MOUSE.H11MO.0.A   0.000545   \n",
       "\n",
       "                  Match_9  q-value.8               Match_10  q-value.9  \\\n",
       "0     JUN_MOUSE.H11MO.0.A   0.004393   JUNB_HUMAN.H11MO.0.A   0.004393   \n",
       "1  RARA_nuclearreceptor_6   0.166742   ZIC3_HUMAN.H11MO.0.B   0.166742   \n",
       "2   FOXO4_HUMAN.H11MO.0.C   0.006791  FOXO4_MOUSE.H11MO.0.C   0.006791   \n",
       "3             JDP2_bZIP_1   0.336541    FOSL1+JUNB_MA1137.1   0.336541   \n",
       "4     SP4_HUMAN.H11MO.0.A   0.000663    SP4_MOUSE.H11MO.0.B   0.000663   \n",
       "\n",
       "               string  \n",
       "0           ATGACTCAT  \n",
       "1      GCGCCCCCTAGTGG  \n",
       "2           TGTTTACAT  \n",
       "3            TTAGTCAT  \n",
       "4  GGCCCCGCCCCCTCCCCG  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_cooperative_net_pos_neg(motif1_only, motif2_only, cooop):\n",
    "    total_sum_th = motif1_only + motif2_only\n",
    "    total_diff_th = np.abs(motif1_only - motif2_only)\n",
    "    total_diff_th1 = np.max([motif1_only,motif2_only])\n",
    "    total_sum_th1 = np.max([motif1_only,motif2_only])\n",
    "    \n",
    "    pos_coop_score = (np.max(cooop) - total_sum_th)/total_sum_th\n",
    "    pos_coop_score1 = (np.max(cooop) - total_sum_th1)/total_sum_th1\n",
    "\n",
    "    neg_coop_score = (np.min(cooop) - total_diff_th)/total_diff_th\n",
    "    neg_coop_score1 = (np.min(cooop) - total_diff_th1)/total_diff_th1\n",
    "        \n",
    "    return pos_coop_score, pos_coop_score1, neg_coop_score, neg_coop_score1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#motifs = [\"TGGAC\",\"TGACTCAT\"]\n",
    "#motifs = [\"AGGAATGT\",\"TTGACTCA\"]\n",
    "values_fresh = {}\n",
    "\n",
    "#for i in range(annotations.shape[0]-1):\n",
    "#    for j in range(i+1,annotations.shape[0]):\n",
    "for i in range(annotations.shape[0]):\n",
    "    for j in range(i,annotations.shape[0]):\n",
    "        label1 = annotations.loc[i,\"Pattern\"] + \"-\" + annotations.loc[j,\"Pattern\"]\n",
    "        label2 = annotations.loc[i,\"Label\"] + \"-\" + annotations.loc[j,\"Label\"]\n",
    "        motifs = [annotations.loc[i,\"string\"],annotations.loc[j,\"string\"]]\n",
    "\n",
    "        #print(label1)\n",
    "        #print(label2)\n",
    "        #print(motifs)\n",
    "        #list_motifs = [[\"CAGGTG\",\"TGACTCAT\"],[\"AGGAATGT\",\"TTGACTCA\"]]\n",
    "        #labels = [\"ZEB-AP1\", \"TEAD-AP1\"]\n",
    "        values_fresh[label1] = [simulate_motif_set(motifs, model, regions_seqs),label2]\n",
    "        #break\n",
    "   #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif =  \"\"\n",
    "batch_size=128\n",
    "full_footprint_control = get_footprint_for_motif(regions_seqs, motif, model, inputlen, batch_size)\n",
    "values_fresh[\"control\"] = full_footprint_control\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('addons/imr90/coop_matrix_all_new_spaced_2.pkl', 'wb') as handle:\n",
    "    pickle.dump(values_fresh, handle)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
